server/sonar-web/src/main/js/app/types.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+export enum BranchType {
+  LONG = 'LONG',
+  SHORT = 'SHORT'
+}
+
+export interface MainBranch {
+  isMain: true;
+  name: undefined;
+  type: BranchType.LONG;
+}
+
+export interface LongLivingBranch {
+  isMain: boolean;
+  name: string;
+  type: BranchType.LONG;
+}
+
+export interface ShortLivingBranch {
+  isMain: false;
+  name: string;
+  status: {
+    bugs: number;
+    codeSmells: number;
+    vulnerabilities: number;
+  };
+  type: BranchType.SHORT;
+}
+
+export type Branch = MainBranch | LongLivingBranch | ShortLivingBranch;
```

Comment:
We can use extend here, but maybe it's not the best exemple since it's a bit longer : 
```
export interface Branch {
  isMain: boolean;
  name?: string;
  type: BranchType;
}

export interface MainBranch extends Branch {
  isMain: true;
  name: undefined;
  type: BranchType.LONG;
}

export interface LongLivingBranch extends Branch {
  name: string;
  type: BranchType.LONG;
}

export interface ShortLivingBranch extends Branch {
  isMain: false;
  name: string;
  status: {
    bugs: number;
    codeSmells: number;
    vulnerabilities: number;
  };
  type: BranchType.SHORT;
}
```